### PR TITLE
Fixed storages not updating after the 'beforeunload' event

### PR DESCRIPTION
**Pull request by Ogurecher**
## Purpose
Storage wrappers (for localStorage and sessionStorage) don't save changes in native storages if these changes take place after the `beforeunload` event. This means that changes that are applied between the `beforeunload` event and the moment the page is closed are discarded and won't be available after the redirect.

This caused an issue with the [msal](https://github.com/AzureAD/microsoft-authentication-library-for-js) library, which uses redirects for authentication and partially clears the sessionStorage between the `beforeunload` and `unload` events. These changes don't persist after redirecting to another page.

## Approach
I added another `saveToNativeStorages` [call](https://github.com/DevExpress/testcafe-hammerhead/compare/master...Ogurecher:msal-fix?expand=1#diff-1940e681e6c3016afc363e9745e22663444618fb5c3f02a14d57085886a9129fR83), which triggers on the `unload` event. In order to do that, I used the methods from [UnloadSandbox](https://github.com/DevExpress/testcafe-hammerhead/compare/master...Ogurecher:msal-fix?expand=1#diff-57d4b3422c7cd2243c2ec8626614685b5a30cad759278196b03ad3777114c5dcR17) class (and refactored them to be usable with events aside from `beforeunload`).

I had to modify some existing tests:
 - [unload-test](https://github.com/DevExpress/testcafe-hammerhead/compare/master...Ogurecher:msal-fix?expand=1#diff-b70d500509337b7e372d85574ab26a152f12c6e300f32b94811d09312fbc0980R3)
 - [attributes-test](https://github.com/DevExpress/testcafe-hammerhead/compare/master...Ogurecher:msal-fix?expand=1#diff-e4f92f4c8ea3d6fdb41dc09e18a2ad235e37b4ebc2d759d61ca5905f50c1df04R586)
 - [storage-test](https://github.com/DevExpress/testcafe-hammerhead/compare/master...Ogurecher:msal-fix?expand=1#diff-349214aaaa75be2700bed887d7c48c44ce73dd06d2aaad91e12ecd4145040d56R251)

In my opinion, we don't need to update storages on the `beforeunload` event anymore, but removing [the listener](https://github.com/DevExpress/testcafe-hammerhead/compare/master...Ogurecher:msal-fix?expand=1#diff-1940e681e6c3016afc363e9745e22663444618fb5c3f02a14d57085886a9129fL74) makes some existing tests fail (they are mostly hard-coded to check the `beforeunload` event). Please let me know your thoughts on this.

## References
Closes https://github.com/DevExpress/testcafe/issues/4834

## Pre-Merge TODO
- [x] Write tests for your proposed changes
- [x] Make sure that existing tests do not fail

### CODE REVIEW

**Review by LavrovArtem — CHANGES_REQUESTED** [review]

File: src/client/sandbox/event/unload.ts [hunk 1]
```diff
@@ -5,23 +5,44 @@ import { isFirefox, isIOS } from '../../utils/browser';
...
+    nativeEventName:   string;
+    eventName:         string;
+    eventPropSetter:   any;
+};
+
 export default class UnloadSandbox extends SandboxBase {
     BEFORE_UNLOAD_EVENT: string = 'hammerhead|event|before-unload';
     BEFORE_BEFORE_UNLOAD_EVENT: string = 'hammerhead|event|before-before-unload';
     UNLOAD_EVENT: string = 'hammerhead|event|unload';
+    NATIVE_UNLOAD_EVENT: string = 'unload';
```
Comment (LavrovArtem) [inline src/client/sandbox/event/unload.ts:21]: It is unnecessary to save the event name, see https://github.com/DevExpress/testcafe-hammerhead/blob/daf68f22dc795cc7d976e5da629eb3ff63175208/src/client/sandbox/event/unload.ts#L51

File: src/client/sandbox/event/unload.ts [hunk 2]
```diff
@@ -30,99 +51,124 @@ export default class UnloadSandbox extends SandboxBase {
...
+        nativeAddEventListener.call(this.window, eventProperties.nativeEventName, this);
     }
 
-    attach (window: Window & typeof globalThis) {
-        super.attach(window);
-
-        this._listeners.setEventListenerWrapper(window, [this.beforeUnloadEventName], (e, listener) => this._onBeforeUnloadHandler(e, listener));
-        this._listeners.addInternalEventListener(window, ['unload'], () => this.emit(this.UNLOAD_EVENT));
+    private _setEventListenerWrapper (eventProperties: EventProperties) {
+        this._listeners.setEventListenerWrapper(window, [eventProperties.nativeEventName], (e, listener) => {
```
Comment (LavrovArtem) [inline src/client/sandbox/event/unload.ts:111]:
```ts
this._listeners.setEventListenerWrapper(window, [eventProperties.nativeEventName], this._createEventHandler(eventProperties));
```

**Review by LavrovArtem — CHANGES_REQUESTED** [review]

File: src/client/sandbox/storages/index.ts [hunk 3]
```diff
@@ -65,13 +67,21 @@ export default class StorageSandbox extends SandboxBase {
...
-                    this.localStorageWrapper.saveToNativeStorage();
-                    this.sessionStorageWrapper.saveToNativeStorage();
+                try {
+                    if (!this.isLocked) {
+                        this.localStorageWrapper.saveToNativeStorage();
+                        this.sessionStorageWrapper.saveToNativeStorage();
+                    }
+                }
+                catch (e) {
+                    if (e.code !== STORAGE_ACCESS_DENIED_ERROR_CODE)
```
Comment (miherlosev) [inline src/client/sandbox/storages/index.ts:77]: Find the cases when this error is occurred and describe them in comment.

File: src/client/sandbox/storages/index.ts [hunk 4]
```diff
@@ -65,13 +67,21 @@ export default class StorageSandbox extends SandboxBase {
...
+                        throw e;
                 }
             };
 
             this._unloadSandbox.on(this._unloadSandbox.BEFORE_UNLOAD_EVENT, saveToNativeStorages);
+            this._unloadSandbox.on(this._unloadSandbox.UNLOAD_EVENT, saveToNativeStorages);
```
Comment (miherlosev) [inline src/client/sandbox/storages/index.ts:83]: Investigate why we need save data to native storages two times.

**Review by miherlosev — CHANGES_REQUESTED** [review]

**Review by LavrovArtem — APPROVED** [review]

**Review by miherlosev — APPROVED** [review]

**Comment by shridharkurhade** [comment]
Hi,
Do we have any complete working example for this. We are facing similar issue while using codeless platform with oidc authentication. Oidc angular library couldn't save oidc config to session storage while running under hammer head and hence login redirect not working.
Please suggest.

**Comment by shridharkurhade** [comment]
Hi,
Do we have any complete working example for this. We are facing similar issue while using codeless platform with oidc authentication. Oidc angular library couldn't save oidc config to session storage while running under hammer head and hence login redirect not working.
Please suggest.

**Comment by Bayheck** [comment]
Please create a separate ticket using this [template](https://github.com/DevExpress/testcafe/issues/new?assignees=&labels=TYPE%3A+bug&template=bug_report.yaml) and also share a [Minimal Working Example](https://testcafe.io/documentation/402636/faq/general-info?search#how-to-create-a-minimal-working-example-when-you-submit-an-issue) in that ticket so that the TestCafe team can research the issue and assist.
 

TestCafe-Hammerhead is a proxy specifically designed for use with TestCafe. Using it separately is not considered a standard use case.

Does your issue occur in [TestCafe](https://github.com/DevExpress/testcafe)?

**Pull request merged**